test/integration/master/audit_dynamic_test.go

```diff
@@ -0,0 +1,279 @@
+		missing, err := testServer2.WaitForEvents(expected[len(expected)-4:], 5*time.Second)
+		require.NoError(t, err, "failed to find the final events in server2, events %#v not found", missing)
+
+		// combine the event lists
+		el1 := testServer1.GetEventList()
+		el2 := testServer2.GetEventList()
+		combinedList := auditinternal.EventList{}
+		combinedList.Items = append(el1.Items, el2.Items...)
+
+		// check that there are no duplicate events
+		dups, err := utils.CheckForDuplicates(combinedList)
+		require.NoError(t, err, "duplicate events found: %#v", dups)
+
+		// check that no events are missing
+		missing, err = utils.CheckAuditList(combinedList, expected)
+		require.NoError(t, err, "failed to match all expected events: %#v not found", missing)
+	})
+}
+
+// sinkHealth checks if sinks are running by verifying that uniquely identified events are found
+// in the given servers
+func sinkHealth(t *testing.T, kubeclient kubernetes.Interface, servers ...*utils.AuditTestServer) {
+	var missing []utils.AuditEvent
+	i := 0
+	var finalErr error
+	err := wait.PollImmediate(50*time.Millisecond, 10*time.Second, func() (bool, error) {
+		i++
+		name := fmt.Sprintf("health-%d-%d", i, time.Now().UnixNano())
+		expected, err := simpleOp(name, kubeclient)
+		require.NoError(t, err, "could not perform config map operations")
+
+		// check that all given servers have received events
+		for _, server := range servers {
+			missing, err = server.WaitForEvents(expected, 5*time.Second)
+			if err != nil {
+				finalErr = fmt.Errorf("not all events found in %s health check: missing %#v", server.Name, missing)
+				return false, nil
+			}
+			server.ResetEventList()
+		}
+		return true, nil
+	})
+	require.NoError(t, err, finalErr)
+}
+
+// simpleOp is a function that simply tries to get a configmap with the given name and returns the
+// corresponding expected audit event
+func simpleOp(name string, kubeclient kubernetes.Interface) ([]utils.AuditEvent, error) {
+	_, err := kubeclient.CoreV1().ConfigMaps(namespace).Get(name, metav1.GetOptions{})
+	if err != nil && !errors.IsNotFound(err) {
+		return nil, err
+	}
+
+	expectedEvents := []utils.AuditEvent{
+		{
+			Level:             auditinternal.LevelRequestResponse,
+			Stage:             auditinternal.StageResponseComplete,
+			RequestURI:        fmt.Sprintf("/api/v1/namespaces/%s/configmaps/%s", namespace, name),
+			Verb:              "get",
+			Code:              404,
+			User:              auditTestUser,
+			Resource:          "configmaps",
+			Namespace:         namespace,
+			RequestObject:     false,
+			ResponseObject:    true,
+			AuthorizeDecision: "allow",
+		},
+	}
+	return expectedEvents, nil
+}
+
+// asyncOps runs the simpleOp function until the stopChan is closed updating
+// the expected atomic events list
+func asyncOps(
+	stopCh <-chan struct{},
+	wg *sync.WaitGroup,
+	kubeclient kubernetes.Interface,
+	expected *atomic.Value,
+) {
+	for i := 0; ; i++ {
```

Comment:
Picking back up the hotlooping thread - if there is an error in simpleOp for some reason, this will hotloop forever. I'm fine with a very short sleep - e.g. 1 millisecond.  Better yet - use a [rate limiter](https://godoc.org/github.com/kubernetes/kubernetes/staging/src/k8s.io/client-go/util/flowcontrol#NewTokenBucketRateLimiter) with sufficient burst to be effectively unthrottled in the happy path, but throttled in the worst case.

Reply:
This client is currently set to `QPS:50 Burst:100`, do we think thats sufficient?

Reply:
Ah, that should be fine. Out of curiosity, where did you find those values?